File: src/main/java/com/wishai/xzrtw/service/ResourceService.java
package com.wishai.xzrtw.service;

import com.wishai.xzrtw.config.XzrtPropertiesFactory;
import com.wishai.xzrtw.exception.FileNotStore;
import com.wishai.xzrtw.model.Applicant;
import com.wishai.xzrtw.model.Article;
import com.wishai.xzrtw.model.ArticleContent;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Map;

public interface ResourceService {

    /* STATIC RESOURCES GOES HERE*/

    String LANG_ZH = "zh";
    String LANG_EN = "en";

    String BLOCK_TYPE_PARAGRAPH = "paragraph";
    String BLOCK_TYPE_IMAGE = "image";

    String CATEGORY_RACE = "race";
    String CATEGORY_NEWS = "news";
    String CATEGORY_TECH = "tech";

    String FILE_TYPE_IMAGE = "image";
    String FILE_TYPE_MUSIC = "music";
    String FILE_TYPE_VIDEO = "video";
    String FILE_TYPE_ZIP = "zip";
    String FILE_TYPE_PDF = "pdf";
    String SRC_TYPE_TEXT = "text";

    String FILE_PATH_IMAGE = "images";
    String FILE_PATH_MUSIC = "musics";
    String FILE_PATH_VIDEO = "videos";
    String FILE_PATH_ZIP = "zips";
    String FILE_PATH_PDF = "pdfs";

    Path FILE_ROOT_PATH = Paths.get(XzrtPropertiesFactory.getXzrtProperties().getUploadRootPath());
    
    static String categoryType2GalleryType(String category) {
        switch (category) {
            case ResourceService.CATEGORY_RACE:
                return "race";
            case ResourceService.CATEGORY_NEWS:
                return "news";
            case ResourceService.CATEGORY_TECH:
                return "tech";
        }
        return null;
    }

    /* get the sub path of file */
    static Path file2Path(String type, String fileName) {
        switch (type) {
            case FILE_TYPE_IMAGE:
                return FILE_ROOT_PATH.resolve(FILE_PATH_IMAGE).resolve(fileName);
            case FILE_TYPE_MUSIC:
                return FILE_ROOT_PATH.resolve(FILE_PATH_MUSIC).resolve(fileName);
            case FILE_TYPE_VIDEO:
                return FILE_ROOT_PATH.resolve(FILE_PATH_VIDEO).resolve(fileName);
            case FILE_TYPE_PDF:
                return FILE_ROOT_PATH.resolve(FILE_PATH_PDF).resolve(fileName);
            case FILE_TYPE_ZIP:
                return FILE_ROOT_PATH.resolve(FILE_PATH_ZIP).resolve(fileName);
        }
        return null;
    }

    /* ?? SHOULD BUILD IN THE IMPL */
    static Article makeArticle(String lang, String author, String category, ArticleContent articleContent) {
        Article article = new Article();

        // initial a article content, and bind to the article
        article.setArticleContentByLang(articleContent, lang);

        // set single non-json data to article
        article.setAuthor(author);
        article.setCategory(category);


        // set the author
        return article;
    }


    /* FUNCTIONS FOR THE ARTICLES */


    ArticleContent makeArticleContent(Map<String, Object> data);

    /* should add the src before create the article */
    Article findArticle(Integer id);

    Integer saveArticle(Article article);

    Integer saveApplicant(Applicant applicant);



    /* FUNCTIONS FOR ADDING AND GETTING SRC */
    Object loadSrc(Integer id, String dType);

    Object loadSrc(Integer id, String dType, boolean isCached);

    Integer storeSrc(Object src, String dType) throws FileNotStore;


    /* FUNCTIONS USED TO TEST, SHOULD BE REMOVED IN THE FUTURE */

    void removeArticle(Integer id);



    /* THINGS ARE NOT REQUIRED ANYMORE */

}

File: src/main/java/com/wishai/xzrtw/service/ResourceService.kt
package com.wishai.xzrtw.service

import com.wishai.xzrtw.config.XzrtPropertiesFactory
import com.wishai.xzrtw.exception.FileNotStore
import com.wishai.xzrtw.model.Applicant
import com.wishai.xzrtw.model.Article
import com.wishai.xzrtw.model.ArticleContent

import java.nio.file.Path
import java.nio.file.Paths

interface ResourceService {

    /* FUNCTIONS FOR THE ARTICLES */
    fun makeArticleContent(data: Map<String, Any>): ArticleContent

    /* should add the src before create the article */
    fun findArticle(id: Int?): Article?

    fun saveArticle(article: Article): Int?

    fun saveApplicant(applicant: Applicant): Int?


    /* FUNCTIONS FOR ADDING AND GETTING SRC */
    fun loadSrc(id: Int, dType: String): Any?

    fun loadSrc(id: Int, dType: String, isCached: Boolean): Any?

    @Throws(FileNotStore::class)
    fun storeSrc(src: Any, dType: String): Int?


    /* FUNCTIONS USED TO TEST, SHOULD BE REMOVED IN THE FUTURE */

    fun removeArticle(id: Int?)

    companion object {

        /* STATIC RESOURCES GOES HERE*/

        val LANG_ZH = "zh"
        val LANG_EN = "en"

        val BLOCK_TYPE_PARAGRAPH = "paragraph"
        val BLOCK_TYPE_IMAGE = "image"

        val CATEGORY_RACE = "race"
        val CATEGORY_NEWS = "news"
        val CATEGORY_TECH = "tech"

        val FILE_TYPE_IMAGE = "image"
        val FILE_TYPE_MUSIC = "music"
        val FILE_TYPE_VIDEO = "video"
        val FILE_TYPE_ZIP = "zip"
        val FILE_TYPE_PDF = "pdf"
        val SRC_TYPE_TEXT = "text"

        val FILE_PATH_IMAGE = "images"
        val FILE_PATH_MUSIC = "musics"
        val FILE_PATH_VIDEO = "videos"
        val FILE_PATH_ZIP = "zips"
        val FILE_PATH_PDF = "pdfs"

//        val FILE_ROOT_PATH = Paths.get(XzrtPropertiesFactory.xzrtProperties!!.uploadRootPath)

        fun categoryType2GalleryType(category: String): String? {
            when (category) {
                ResourceService.CATEGORY_RACE -> return "race"
                ResourceService.CATEGORY_NEWS -> return "news"
                ResourceService.CATEGORY_TECH -> return "tech"
            }
            return null
        }

        /* get the sub path of file */
        fun file2Path(type: String, fileName: String): Path? {
            val FILE_ROOT_PATH = Paths.get(XzrtPropertiesFactory.xzrtProperties!!.uploadRootPath)
            when (type) {
                FILE_TYPE_IMAGE -> return FILE_ROOT_PATH.resolve(FILE_PATH_IMAGE).resolve(fileName)
                FILE_TYPE_MUSIC -> return FILE_ROOT_PATH.resolve(FILE_PATH_MUSIC).resolve(fileName)
                FILE_TYPE_VIDEO -> return FILE_ROOT_PATH.resolve(FILE_PATH_VIDEO).resolve(fileName)
                FILE_TYPE_PDF -> return FILE_ROOT_PATH.resolve(FILE_PATH_PDF).resolve(fileName)
                FILE_TYPE_ZIP -> return FILE_ROOT_PATH.resolve(FILE_PATH_ZIP).resolve(fileName)
            }
            return null
        }

        /* ?? SHOULD BUILD IN THE IMPL */
        fun makeArticle(lang: String, author: String, category: String, articleContent: ArticleContent): Article {
            val article = Article()

            // initial a article content, and bind to the article
            article.setArticleContentByLang(articleContent, lang)

            // set single non-json data to article
            article.author = author
            article.category = category

            // set the author
            return article
        }
    }


    /* THINGS ARE NOT REQUIRED ANYMORE */

}
